### PR TITLE
Enable Cargo's sparse protocol in CI

### LINKED ISSUE #107342: Dogfood `CARGO_REGISTRIES_CRATES_IO_PROTOCOL=sparse` in CI
Sparse-registry was recently stabilized and cargo itself has started dogfooding it in CI: https://github.com/rust-lang/cargo/pull/11632. We should do the same in rust-lang/rust, both to speed up CI and to test out the feature.

I think around https://github.com/rust-lang/rust/blob/beefcf88544b7f4d1b5e9ae77f93566303d34326/src/ci/run.sh#L129-L130 is the right place to set this; we should verify by inspection that the feature is being used in CI before merging.

### PR DESCRIPTION
**Pull request by ehuss**
This enables the sparse protocol in CI in order to exercise and dogfood it. This is intended test the production server in a real-world situation.

Closes #107342

### CODE REVIEW

File: src/ci/run.sh [hunk 1]
```diff
@@ -45,6 +45,8 @@ fi
 ci_dir=`cd $(dirname $0) && pwd`
 source "$ci_dir/shared.sh"
 
+export CARGO_REGISTRIES_CRATES_IO_PROTOCOL=sparse
```
Comment (jyn514) [inline src/ci/run.sh:48]: Hmm, why does this need to be set both here and in the `ci.yml` file? I would expect it to only need to be set here.

**Review by jyn514 — APPROVED** [review]
Thanks! r=me with or without the nit addressed :)

File: src/ci/run.sh (hunk 1 above)
Comment (ehuss) [reply src/ci/run.sh:48]: The environment isn't inherited inside Docker images. I could use the `--env` flag to forward it, but I felt less confident that would cover everything.

**Comment by jyn514** [comment]
Hmm, I still see "updating crates.io index" in the build logs. Is that expected?

https://github.com/rust-lang/rust/actions/runs/4060238532/jobs/6989128411#step:26:1096

**Comment by ehuss** [comment]
Yes. With the sparse protocol, Cargo still needs to download registry index files covering the dependencies that are needed. That's what the `Updating crates.io index` line is referring to.

**Comment by jyn514** [comment]
👍 

@bors r+ rollup=iffy

**Pull request merged**